[IT-953] & [IT-958] Refactor to use Pipenv; Publish to AWS Serverless Repo

boyleconnor opened the pull request.

Commit:
Refactor to use Pipenv; Publish to AWS Serverless Repo

Commit:
Fix Formatting on template.yaml

Commit:
Fix Formatting on .travis.yml

xschildw reviewed (approved):
Which shell are you running in travis? I tried that syntax to build files with echo on Jenkins and '\n' always came back as '\n' strings (not 0x0A)...

boyleconnor commented:
> Which shell are you running in travis? I tried that syntax to build files with echo on Jenkins and '\n' always came back as '\n' strings (not 0x0A)...

@xschildw as far as I'm aware it's set to whatever the default is for Travis (Bash I guess?)

Are you making sure to include the `-e` option on `echo`? The default behavior otherwise is to not expand backslash escapes.

zaro0508 commented on .travis.yml:
```diff
@@ -2,14 +2,22 @@ language: python
+    - REPO_NAME="${PWD##*/}"
+    - PIPENV_IGNORE_VIRTUALENVS=1
+    - SAM_CLI_TELEMETRY=0
+    - AWS_DEFAULT_REGION="us-east-1"
```
these AWS env vars are not needed.  AWS configs are setup in `before_script` section.  it should look like our standard lambda template: https://github.com/Sage-Bionetworks-IT/lambda-template/blob/master/.travis.yml

zaro0508 commented on Pipfile:
```diff
@@ -13,6 +13,7 @@ pre-commit = "*"
 aws-sam-cli = "*"
 boto3 = "*"
 botocore = "*"
+jsondiff = "*"
```
all dependencies should contain versions, like in our standard template https://github.com/Sage-Bionetworks-IT/lambda-template/blob/master/Pipfile

also please put dependencies in correct sections `dev-package` are only needed for building and testing the lambda.  in this case i think all libs should go in `dev-packages` and remove `packages` section.

you should also commit the corresponding `Pipfile.lock` file.  Run `pipenv install --dev` to get an updated Pipfile.lock then commit that.

zaro0508 commented on template.yaml:
```diff
@@ -41,3 +41,17 @@ Resources:
+    ReadmeUrl: "README.md"
+    Labels: ["serverless", "template"]
+    HomePageUrl: "https://github.com/Sage-Bionetworks-IT/cfn-explode-macro"
+    SemanticVersion: "master"
```
must be a semantic version number.  change this to `0.0.1` will work.

zaro0508 commented on template.yaml:
```diff
@@ -41,3 +41,17 @@ Resources:
+    Labels: ["serverless", "template"]
+    HomePageUrl: "https://github.com/Sage-Bionetworks-IT/cfn-explode-macro"
+    SemanticVersion: "master"
+    SourceCodeUrl: "https://github.com/Sage-Bionetworks-IT/cfn-explode-macro/tree/master"
```
master/0.0.1

zaro0508 commented on template.yaml:
```diff
@@ -41,3 +41,17 @@ Resources:
+  AWS::ServerlessRepo::Application:
+    Name: "cfn-explode-macro"
+    Description: "macro that provides a template-wide Explode property for CloudFormation resources"
+    Author: "zaro0508"
```
lets set this to Sage-Bionetworks

zaro0508 commented on template.yaml:
```diff
@@ -41,3 +41,17 @@ Resources:
+    # paths are relative to .aws-sam/build directory
+    LicenseUrl: "LICENSE"
+    ReadmeUrl: "README.md"
+    Labels: ["serverless", "template"]
```
updates with labels specific to this lambda.  maybe add `serverless`, `macro`, and `cloudformation`

zaro0508 commented on template.yaml:
```diff
@@ -41,3 +41,17 @@ Resources:
+    Description: "macro that provides a template-wide Explode property for CloudFormation resources"
+    Author: "zaro0508"
+    SpdxLicenseId: "Apache-2.0"
+    # paths are relative to .aws-sam/build directory
```
remove comment

tthyer reviewed (changes requested):
@cascadianblue I'm a little conflicted about this one. Definitely like finishing the switch to pipenv. But, the lambda doesn't need awscli, that's only for the CI, plus you have to do all those settings of environment variables and creating the `.aws` files. I advocate creating a new version of `setup_aws_cli.sh` that doesn't append the TRAVIS_BRANCH to the environment variables (which causes problems) instead -- is that why you were motivated to change it? I have had problems with that script when attempting to build development branches in CI.

That's my $0.02. What does @zaro0508 think?

zaro0508 commented:
> @cascadianblue I'm a little conflicted about this one. Definitely like finishing the switch to pipenv. But, the lambda doesn't need awscli, that's only for the CI, plus you have to do all those settings of environment variables and creating the `.aws` files. I advocate creating a new version of `setup_aws_cli.sh` that doesn't append the TRAVIS_BRANCH to the environment variables (which causes problems) instead -- is that why you were motivated to change it? I have had problems with that script when attempting to build development branches in CI.
> 
> That's my $0.02. What does @zaro0508 think?

I've asked @cascadianblue to make these changes @tthyer.  i've instructed to remove `packages` section in Pipfile.  Also the `before_script` section in .travis.yml replaces `setup_aws_cli.sh` script.

Commit:
Remove Unneeded AWS Env Vars

Commit:
Mobe Development Packages to Correct Section of Pipfile

Commit:
Add Updated Pipfile.lock

Commit:
Add Package Versions to Pipfile

Commit:
Fix Metadata

zaro0508 commented on Pipfile:
```diff
@@ -4,15 +4,14 @@ url = "https://pypi.org/simple"
+botocore = "~=1.17.58"
+jsondiff = "~=1.2.0"
+pre-commit = "~=2.7.1"
+pylint = "~=2.6.0"
```
don't need this because we are using pre-commit

boyleconnor replied:
fixed

zaro0508 replied:
pylint is not needed, please remove.

zaro0508 commented on Pipfile:
```diff
@@ -4,15 +4,14 @@ url = "https://pypi.org/simple"
-boto3 = "*"
-botocore = "*"
+aws-sam-cli = "~=1.2.0"
+boto3 = "~=1.14.58"
```
need to add awscli to this list and we don't need boto3 or botocore because those will get brought in by awscli & aws-sam-cli

boyleconnor replied:
fixed

zaro0508 commented on Pipfile:
```diff
@@ -4,15 +4,14 @@ url = "https://pypi.org/simple"
+pre-commit = "~=2.7.1"
+pylint = "~=2.6.0"
+pytest = "~=6.0.1"
+pytest-mock = "~=3.3.1"
```
need to add awscli to this list

zaro0508 commented on template.yaml:
```diff
@@ -41,3 +41,16 @@ Resources:
+    Labels: ["serverless", "macro", "cloudformation"]
+    HomePageUrl: "https://github.com/Sage-Bionetworks-IT/cfn-explode-macro"
+    SemanticVersion: "0.0.1"
+    SourceCodeUrl: "https://github.com/Sage-Bionetworks-IT/cfn-explode-macro/tree/master/0.0.1"
```
remove `master`

boyleconnor replied:
fixed

Commit:
Update Dependencies & Dependency Versions

Commit:
Fix Typo in template.yaml

Commit:
Remove pylint from Build Reqs

Commit:
Try to Fix Pipfile.lock

It might have gotten broken from uninstalling pylint?

Commit:
Manually Install Importlib-Metadata

Commit:
Add importlib-resources

zaro0508 commented on Pipfile:
```diff
@@ -4,15 +4,14 @@ url = "https://pypi.org/simple"
-botocore = "*"
+awscli = "~=1.18.136"
+aws-sam-cli = "~=1.2.0"
+importlib-metadata = "~=1.7.0"
```
why is `importlib-*` needed?

boyleconnor replied:
Hmmm I hadn't read the error messages carefully and I assumed that something in our code was relying on it. It actually looks like pre-commit is [relying on importlib-resources](https://travis-ci.com/github/Sage-Bionetworks-IT/cfn-explode-macro/builds/184261785), but it doesn't get installed as a dependency for some reason

boyleconnor replied:
Also, SAM CLI seems to [depend on importlib-metadata](https://travis-ci.com/github/Sage-Bionetworks-IT/cfn-explode-macro/builds/184251402#L243).

boyleconnor replied:
Looks like others have faced a similar problem: https://github.com/pre-commit/pre-commit/issues/853

How about I try removing these dependencies and updating the Python version to 3.8 ?

tthyer replied:
I had the same problem with pre-comit and solved it by bumping python from 3.6 to 3.7 in .travis.yml (in the lambda-budgets repo)

boyleconnor replied:
Didn't work

boyleconnor replied:
@zaro0508 could we just leave the importlib-* imports in? It seems to be the least kludge-y way of fixing this problem.

Commit:
Try to Fix importlib Dependency Issues

SAM CLI and Pre-Commit seem to depend on importlib-metadata and
importlib, but they aren't getting installed by default. Maybe changing
the Python version will change that?

Commit:
Revert "Try to Fix importlib Dependency Issues"

This reverts commit a5186b90c8d813dc3af01ac281410414e2775916.

Commit:
Try Python Version 3.7 with no Importlib Install

This reverts commit c77b1caff4ccd0a2b0ce1312ebd4130ff163720c.

Commit:
Pin Version of pre-commit

Commit:
Revert "Pin Version of pre-commit"

This reverts commit 0a2c4d661a6637d9464996c640224ddd89fa2fb7.

Revert "Try Python Version 3.7 with no Importlib Install"

This reverts commit 0a496329442aa2596aef37cc9c0575da2e15f199.

Revert "Revert "Try to Fix importlib Dependency Issues""

This reverts commit c77b1caff4ccd0a2b0ce1312ebd4130ff163720c.

Revert "Try to Fix importlib Dependency Issues"

This reverts commit a5186b90c8d813dc3af01ac281410414e2775916.

Commit:
Set Python Version to 3.7

Commit:
Update Python Version Across Board

Commit:
Make awscli Version Match lambda-template

Commit:
Rebuild Pipfile.lock w/ Dependencies Removed in my Env

Commit:
No Pipfile.lock

Commit:
Set Python Version to 3.8

Commit:
Let's See if this Breaks the Build

Commit:
Revert "Let's See if this Breaks the Build"

This reverts commit 4c42088a6fa6961cd545c1cc879c67c339451249.

Commit:
Switch to Python 3.7 & RM Pipfile.lock

Parts of my local (virtual) environment might be somehow rubbing off
through Pipfile.lock.
